[ASViewController] Add support for ASTraitCollection

rcancro opened the pull request:
Initial attempt to get display traits working with ASEnvironment.

To get proper ASDisplayTraits support, you must use an ASViewController. The ASViewController implements UITraitCollection-related methods (`traitCollectionDidChange:`, `willTransitionToTraitCollection:withTransitionCoordinator:`, `viewWillTransitionToSize:withTransitionCoordinator`) to update the internal ASDisplayTraits and propagate them to subnodes.

ASTableNode and ASCollectionNode don't actually have their cells as subnodes, so a little bit of trickery is involved (on `setEnvironment:` the table/collection node gets its data controller's completedNodes and propagates the new traits. see `ASDisplayTraitsCollectionTableSetEnvironmentState`). The data controller also has the ability to get the current display traits when creating new cells through a new protocol `ASDataControllerEnvironmentDelegate`.

ASViewController also supports the ability to return a custom set of display traits regardless its `UITraitCollection`. For example, if you have a modal dialog that should always have a compact size classes, you can set the override block before displaying the VC (`overrideDisplayTraitsWithTraitCollection` and `overrideDisplayTraitsWithWindowSize`).

A new example, called Display Traits, has been added. It shows how display traits can be used in a ASViewController with a normal ASDisplayNode as its root, as well as in ASViewControllers hosting a table node or collection node. There is also an example of overriding the default display traits of a VC.

Please provide any suggestions that you have!

maicki commented on AsyncDisplayKit/Details/ASEnvironment.h:
```diff
@@ -93,6 +128,33 @@ ASDISPLAYNODE_EXTERN_C_END
+// This logic is used in both ASCollectionNode and ASTableNode
+#define ASDisplayTraitsCollectionTableSetEnvironmentState \
+- (void)setEnvironmentState:(ASEnvironmentState)environmentState\
+{\
```
Do we need locking for setting and propagation?

maicki commented on examples/DisplayTraits/Sample/CollectionViewController.m:
```diff
@@ -0,0 +1,72 @@
+
+- (ASSizeRange)collectionView:(ASCollectionView *)collectionView constrainedSizeForNodeAtIndexPath:(NSIndexPath *)indexPath
+{
+  ASDisplayTraits displayTraits = self.collectionNode.environmentState.displayTraits;
```
Maybe we should provide a better API to get to the displayTraits so the consumer does not need to have to know (and I think should not know) about the environmentState

cc @appleguy

maicki commented on examples/DisplayTraits/Sample/KittenNode.m:
```diff
@@ -0,0 +1,170 @@
+      .verticalSizeClass = UIUserInterfaceSizeClassCompact,
+      .forceTouchCapability = traitCollection.forceTouchCapability,
+    };
+  };
```
Same for here. Seems pretty difficult for API consumer that are not so familiar with structs or so to use the API. Why about creating an object that basically is a representation for a ASDisplayTraits struct and that will be returned in that block. In ASDK if you call the block we use that object and create a ASDisplayTrait for it. Thoughts?

cc @appleguy

maicki commented on AsyncDisplayKit/Private/ASEnvironmentInternal.mm:
```diff
@@ -15,11 +15,17 @@
+BOOL ASEnvironmentStateDownwardPropagationEnabled()
+{
+  return AS_SUPPORT_DOWNWARD_PROPAGATION;
 }
```
As we are currently prevent certain state to propagate should we be more discrete what we support for now? Something like -supportsTraitsCollectionPropagation and ASEnvironmentStateTraitCollectionPropagationEnabled? Maybe we can add that to special header file that if we finally decide to support upward and downward propagation for all states we can just remove it?

maicki commented:
@rcancro  Overall seems good to me. Just one thing I'm thinking about is that the naming of the ASDisplayTraits is a bit different as all the other states that we start with "ASEnvironment..." Just thinking if we should internally use ASEnvironmentDisplayTraitsState and for the public API we expose it as displayTraits as we already do?

 cc @appleguy

rcancro commented:
@maicki Thanks for the review! I've addressed most of your comments except for the request to make a NSObject-based class that wraps the struct. I'll look at that tomorrow

levi commented on AsyncDisplayKit/ASViewController.mm:
```diff
@@ -132,4 +142,70 @@ - (ASInterfaceState)interfaceState
+  [self progagateNewDisplayTraits:displayTraits];
+}
+
+- (void)viewWillTransitionToSize:(CGSize)size withTransitionCoordinator:(id<UIViewControllerTransitionCoordinator>)coordinator
```
Why do we need to implement all three of these notifications? `willTransitionToTraitCollection:` should suffice, as it is called in `transitionToSize:` call and is invoked before `traitCollectionDidChange:` is called. I'd love to learn more about the reasons behind this implementation.

rcancro replied:
I noticed that on start up of the app, `traitCollectionDidChange:` is called, but not `willTransitionToTraitCollection:`.

I also feel like i've seen the case where a window changes size (but not trait collection) and `viewWillTransitionToSize:` is called but not `willTransitionToTraitCollection:`.

`progagateNewDisplayTraits:` does check to make sure that the traits are different before propagating, so even in the case where all three of these are called we only propagate once.

appleguy commented:
@rcancro I am re-running, but I think this may be a real failure.  Probably worth opening the CarthageBuildTest to run it locally and see what the detailed failure is, and confirm the build server's result that before this change, it builds.  It is possible that there is a new problem in master that hasn't been noticed yet though, as we did just adopt a new PINRemoteImage which apparently has some issue with Carthage.  cc @garrettmoon 

▸ Building Sample/CarthageExample [Debug]
▸ Check Dependencies
▸ Compiling main.m
▸ Compiling AppDelegate.m
❌  /Users/travis/build/facebook/AsyncDisplayKit/examples/CarthageBuildTest/CarthageExample/AppDelegate.m:9:9: could not build module 'AsyncDisplayKit'
@import AsyncDisplayKit;

appleguy commented:
@rcancro @levi if ASDisplayTraits is directly analogous to UITraitCollection, can we call it ASTraitCollection instead?  I assume it will be a struct directly inside ASEnvironment struct.

Sincere apologies that I have not been able to review this yet.  I really badly want to, but haven't been able to reach the minimum sleep quota recently even without touching ASDK diffs in the last week :(.  Will plan to review on the plane on Friday or on return flight.

appleguy commented on examples/BackgroundPropertySetting/Sample/ViewController.swift:
```diff
@@ -24,8 +24,8 @@ final class ViewController: ASViewController, ASCollectionDelegate, ASCollection
 		super.init(node: ASCollectionNode(collectionViewLayout: layout))
-		navigationItem.rightBarButtonItem = UIBarButtonItem(title: "Color", style: .Plain, target: self, action: "didTapColorsButton")
-		navigationItem.leftBarButtonItem = UIBarButtonItem(title: "Layout", style: .Plain, target: self, action: "didTapLayoutButton")
+        navigationItem.rightBarButtonItem = UIBarButtonItem(title: "Color", style: .Plain, target: self, action: #selector(didTapColorsButton))
```
Needs more indentation?

levi commented on .editorconfig:
```diff
@@ -0,0 +1,19 @@
+# http://editorconfig.org
```
Include this file in a separate patch.

levi commented:
@rcancro looks like this patch got messed up with a rebase or merge.

levi commented on AsyncDisplayKit/ASViewController.mm:
```diff
@@ -132,4 +142,70 @@ - (ASInterfaceState)interfaceState
+  
+  if (ASEnvironmentTraitCollectionIsEqualToASEnvironmentTraitCollection(traitCollection, oldTraitCollection) == NO) {
+    environmentState.traitCollection = traitCollection;
+    [self.node setEnvironmentState:environmentState];
```
Are you able to use dot syntax here?

rcancro replied:
you mean `self.node.environmentState = environmentState`?

levi replied:
yeah

levi commented on AsyncDisplayKit/ASViewController.mm:
```diff
@@ -132,4 +142,70 @@ - (ASInterfaceState)interfaceState
+  return self.node.environmentState.traitCollection;
+}
+
+- (void)progagateNewDisplayTraits:(ASEnvironmentTraitCollection)traitCollection
```
Is there any value in moving this to the node itself? Seems like it doesn't have a lot of relationship with the view controller.

rcancro replied:
The thinking was that the VC would be the one that knows of any special trait collection it may need based on the circumstances. Like UIViewController has  setOverrideTraitCollection method, not UIView.

rcancro replied:
Oh, you are talking about moving `progagateNewDisplayTraits:` to the node level. That is possible. The VC will still be the one to initiate the propagation since it is the one that knows when the trait collection actually changed. Personally, I like that all the trait collection details are encapsulated inside of the VC, but I definitely see your point.

levi commented on AsyncDisplayKit/ASViewController.mm:
```diff
@@ -132,4 +142,70 @@ - (ASInterfaceState)interfaceState
+    traitCollection.isMutable = NO;
+    return [traitCollection environmentTraitCollection];
+  }
+  return self.node.environmentState.traitCollection;
```
Would be nice to reduce the amount of references here. Perhaps add a shortcut to asdisplaynode to get the underlying environmentState trait collection?

rcancro replied:
@maicki had the idea to expose the trait collection as ASTraitCollection to the user, but store it as a struct behind the scenes. When providing a custom trait collection the user overrides one of two blocks, both of which return ASTraitCollection* not the struct.

I guess my point is that we are trying to make it difficult for the user to get to the struct trait collection since most obj-c developers are more familiar using classes. By adding a getter to the struct traits we kind of break that.

levi replied:
We could introduce it as a private method for these use cases.

levi commented on AsyncDisplayKit/ASViewController.mm:
```diff
@@ -132,4 +142,70 @@ - (ASInterfaceState)interfaceState
 
+#pragma mark - ASEnvironmentTraitCollection
+
+- (ASEnvironmentTraitCollection)displayTraitsForTraitCollection:(UITraitCollection *)traitCollection
```
Might be a better opportunity to allow the developer to provide a subclass of ASTraitCollection (or ASEnvironmentTraitCollection) with extra properties that `displayTraitsContext` currently provide. For the common case, most people won't need to have `displayTraitsContext`, so figuring out a way to reduce API footprint would be a bonus. Are we constrained introducing a design like that?

rcancro replied:
Ultimately we are passing around trait collections as a struct, not a class. It has to be a struct to live in ASEnvironment. The class is just there to be a nicer API to the end user (that probably doesn't care about the context).

levi commented on AsyncDisplayKit/Details/ASTraitCollection.h:
```diff
@@ -0,0 +1,29 @@
+
+@interface ASTraitCollection : NSObject
+
+@property (nonatomic, assign) BOOL isMutable;
```
I'd recommend dropping this in favor of making the ASTraitCollection immutable the moment it's initialized. Will require a larger initializer, but will reduce potential mistakes.

rcancro replied:
Sounds good. FWIW, when ASViewController creates a ASTraitCollection via its override blocks, it immediately sets it to immutable (i.e., the user doesn't have to do it). This is similar to how layout specs get set to immutable. 

Though I'm not wild that isMutable is read/write so a user could just set it back to NO.

levi commented on AsyncDisplayKit/ASDisplayNode.mm:
```diff
@@ -2710,9 +2713,19 @@ - (BOOL)supportsUpwardPropagation
 ASEnvironmentLayoutOptionsForwarding
 ASEnvironmentLayoutExtensibilityForwarding
 
+- (ASTraitCollection *)asyncTraitCollection
```
What about asdk_traitCollection?

rcancro replied:
![screen shot 2016-05-10 at 16 33 57](https://cloud.githubusercontent.com/assets/285809/15165735/04a5b16a-16cd-11e6-8159-d3ceec5bd15d.png)

levi replied:
-_-

levi commented on AsyncDisplayKit/Details/ASEnvironment.mm:
```diff
@@ -0,0 +1,83 @@
+  return asyncTraitCollection;
+}
+
+BOOL ASEnvironmentTraitCollectionIsEqualToASEnvironmentTraitCollection(ASEnvironmentTraitCollection traitCollection0, ASEnvironmentTraitCollection traitCollection1)
```
lhs, rhs instead of traitCollection0, traitCollection1

levi commented on AsyncDisplayKit/Details/ASDataController.mm:
```diff
@@ -519,8 +520,17 @@ - (void)accessDataSourceSynchronously:(BOOL)synchronously withBlock:(dispatch_bl
+      ASCellNodeBlock nodeBlockPropagatingDisplayTraits = ^{
+        ASCellNode *cellNode = nodeBlock();
+        id<ASEnvironment> environment = [self.environmentDelegate dataControllerEnvironment];
+        ASEnvironmentStatePropagateDown(cellNode, [environment environmentState].traitCollection);
```
would be nice to have something like `environment.traitCollection`, even if it's brought in by a category.

levi commented on AsyncDisplayKit/ASViewController.h:
```diff
@@ -9,14 +9,31 @@
 
 @property (nonatomic, strong, readonly) DisplayNodeType node;
 
+@property (nonatomic, strong) id displayTraitsContext;
```
The documentation of this property should be very clear — its usefulness and also the internal memory behavior.

levi commented on AsyncDisplayKit/ASViewController.h:
```diff
@@ -22,7 +22,17 @@ typedef ASTraitCollection * _Nonnull (^ASDisplayTraitsForTraitWindowSizeBlock)(C
+ *  ASVC keeps a strong reference to the context to make sure that it stays alive. If you change this value
+ *  it will propagate the change to the subnodes.
+ */
+@property (nonatomic, strong) id _Nullable traitColectionContext;
```
typo

rcancro replied:
d'oh! good catch

levi commented on AsyncDisplayKit/ASViewController.mm:
```diff
@@ -50,9 +51,8 @@ - (instancetype)initWithNode:(ASDisplayNode *)node
-  if (_displayTraitsContext != nil) {
-    ASEnvironmentTraitCollectionClearDisplayContext(self.node);
-    _displayTraitsContext = nil;
+  if (_traitColectionContext != nil) {
```
Add a note as to why you're clearing it out in dealloc.

levi commented:
Ricky, this is straight up poetry.

rcancro commented:
Rotation, oh my
iPad, 6 plus and pad pro
trait collections yay!

appleguy commented on AsyncDisplayKit/Details/ASEnvironment.h:
```diff
@@ -93,6 +127,41 @@ ASDISPLAYNODE_EXTERN_C_END
+#define ASEnvironmentCollectionTableSetEnvironmentState(lock) \
+- (void)setEnvironmentState:(ASEnvironmentState)environmentState\
+{\
+  ASDN::MutexLocker l(lock);\
```
Pretty damn fancy, sir @rcancro

appleguy commented on AsyncDisplayKit/Layout/ASLayoutSpec.mm:
```diff
@@ -168,11 +169,13 @@ - (NSArray *)children
 
 - (ASEnvironmentState)environmentState
 {
+  ASDN::MutexLocker l(_propertyLock);
```
I am not certain this is safe.  I believe there is a deadlock risk here - let's reference the blame in this file for the change that had to remove a certain lock.  @rcancro @garrettmoon @maicki

maicki replied:
Agree we have to be careful with adding locks in ASEnvironment. E.g. we didn't lock layout options in the past before ASEnvironment and as we added locks for the ASEnvironmentLayoutOptions deadlocks happened. Especially if we traverse the node tree's from bottom up (e.g. it happens if called `setNeedsLayout`) and the same time the layout happens from top down (e.g. call `measureWithSizeRange:` directly)

appleguy commented on AsyncDisplayKit/Layout/ASLayoutSpec.mm:
```diff
@@ -193,9 +201,20 @@ - (void)propagateUpLayoutable:(id<ASLayoutable>)layoutable
 
+- (ASEnvironmentTraitCollection)environmentTraitCollection
+{
+  return _environmentState.traitCollection;
```
I think the locking may be unsafe in this file, BUT - if we do lock for accessing _environmentState below in asyncTraitCollection, we probably should lock here

appleguy commented:
Phenomenal work, @rcancro !  Once again, you've fundamentally extended the capabilities of the framework.  I sincerely appreciate your attention to detail both in architecting and executing this.  There's great care to efficiency / performance, but simultaneously a clear priority placed on a friendly object-based API.  Love it.

The only followup that seems important for the next 1-2 days is checking on the locking sites I commented at, but otherwise I think this is a mic drop. :)
